vortexor/src/load_balancer.rs

```diff
@@ -0,0 +1,206 @@
+            Arc,
+        },
+        thread::{self, sleep, Builder, JoinHandle},
+        time::{Duration, Instant},
+    },
+};
+
+/// The RpcLoadBalancer can support providing a RpcClient which has the most
+/// up-to-date slot information among a set of configured RPC servers.
+/// This service starts threads which will subscribe to the slot events
+/// using the corresponding web socket for a RPC server. This mechansim can
+/// load balance the RPC calls to multiple RPC servers.
+pub struct RpcLoadBalancer {
+    /// (ws_url, slot)
+    server_to_slot: Arc<DashMap<String, Slot>>,
+    /// (rpc_url, client)
+    server_to_rpc_client: DashMap<String, Arc<RpcClient>>,
+    subscription_threads: Vec<JoinHandle<()>>,
+}
+
+impl RpcLoadBalancer {
+    const DISCONNECT_WEBSOCKET_TIMEOUT: Duration = Duration::from_secs(30);
+    const RPC_TIMEOUT: Duration = Duration::from_secs(15);
+    pub const SLOT_QUEUE_CAPACITY: usize = 100;
+
+    /// Start a Load Balancer service given a list of RPC servers. Each RPC server
+    /// must have the (RPC URL and corresponding Websocket URL) set.
+    /// This create the RpcClients and create web socket connections to the corresponding
+    /// RPC server and watch for the slot events.
+    pub fn new(
+        servers: &[(String, String)], /* http rpc url, ws url */
+        exit: &Arc<AtomicBool>,
+    ) -> (RpcLoadBalancer, Receiver<Slot>) {
+        let server_to_slot = Arc::new(DashMap::from_iter(
+            servers.iter().map(|(_, ws)| (ws.clone(), 0)),
+        ));
+
+        let server_to_rpc_client = DashMap::from_iter(servers.iter().map(|(rpc_url, ws)| {
+            // warm up the connection
+            let rpc_client = Arc::new(RpcClient::new_with_timeout_and_commitment(
+                rpc_url,
+                Self::RPC_TIMEOUT,
+                CommitmentConfig {
+                    commitment: CommitmentLevel::Processed,
+                },
+            ));
+            if let Err(e) = rpc_client.get_slot() {
+                error!("error warming up rpc: {rpc_url}. error: {e}");
+            }
+            // store as ws instead of http so we can lookup by furthest ahead ws subscription
+            (ws.clone(), rpc_client)
+        }));
+
+        // sender tracked as health_manager-channel_stats.slot_sender_len
+        let (slot_sender, slot_receiver) = crossbeam_channel::bounded(Self::SLOT_QUEUE_CAPACITY);
+        let subscription_threads =
+            Self::start_subscription_threads(servers, server_to_slot.clone(), slot_sender, exit);
+        (
+            RpcLoadBalancer {
+                server_to_slot,
+                server_to_rpc_client,
+                subscription_threads,
+            },
+            slot_receiver,
+        )
+    }
+
+    /// Start threads to watch for slot events on the RPC servers and update
+    /// the server_to_slot map.
+    fn start_subscription_threads(
+        servers: &[(String, String)],
```

Comment:
String here is probably not the best type. Maybe parse before calling this function so types explain what those strings should be?

Reply:
changed to Url